engine/verification/fetcher/engine.go

```diff
@@ -106,352 +102,210 @@ func (e *Engine) Done() <-chan struct{} {
 	lg := e.log.With().
-		Hex("chunk", chunkID[:]).
-		Hex("block", blockID[:]).
-		Hex("result_id", resultID[:]).
+		Hex("result_id", logging.ID(locator.ResultID)).
+		Uint64("chunk_index", locator.Index).
 		Logger()
+	lg.Info().Msg("chunk locator arrived")
 
-	sealed, header, err := blockIsSealed(e.state, e.headers, blockID)
-
+	// retrieves result and chunk using the locator
+	result, err := e.results.ByID(locator.ResultID)
 	if err != nil {
-		lg.Error().Err(err).Msg("could not check if block is sealed")
-		e.chunkConsumerNotifier.Notify(chunkID)
-		return
+		lg.Fatal().Err(err).Msg("could not retrieve result for chunk locator")
 	}
+	chunk := result.Chunks[locator.Index]
+	chunkID := chunk.ID()
 
-	// skip sealed blocks
-	if sealed {
-		lg.Debug().Msg("skip sealed chunk")
-		e.chunkConsumerNotifier.Notify(chunkID)
-		return
-	}
-
-	lg = lg.With().Uint64("height", header.Height).Logger()
-
-	err = e.processChunk(c, header, resultID)
+	lg = lg.With().
+		Hex("chunk_id", logging.ID(chunkID)).
+		Hex("block_id", logging.ID(chunk.ChunkBody.BlockID)).
+		Logger()
+	lg.Debug().Msg("result and chunk for locator retrieved")
 
+	// skips processing a chunk if it belongs to a sealed block.
+	sealed, err := e.blockIsSealed(chunk.ChunkBody.BlockID)
 	if err != nil {
-		lg.Error().Err(err).Msg("could not process chunk")
-		// we report finish processing this chunk even if it failed
-		e.chunkConsumerNotifier.Notify(chunkID)
-	} else {
-		lg.Info().Msgf("processing chunk")
+		lg.Fatal().Err(err).Msg("could not determine whether block has been sealed")
 	}
-}
-
-func (e *Engine) processChunk(c *flow.Chunk, header *flow.Header, resultID flow.Identifier) error {
-	blockID := c.ChunkBody.BlockID
-	receiptsData, err := e.receiptsDB.ByBlockID(blockID)
-	if err != nil {
-		return fmt.Errorf("could not retrieve receipts for block: %v: %w", blockID, err)
+	if sealed {
+		e.chunkConsumerNotifier.Notify(chunkID) // tells consumer that we are done with this chunk.
+		lg.Info().Msg("drops requesting chunk of a sealed block")
+		return
 	}
 
-	var receipts []*flow.ExecutionReceipt
-	copy(receipts, receiptsData)
-
-	agrees, disagrees := executorsOf(receipts, resultID)
-	// chunk data pack request will only be sent to executors who produced the same result,
-	// never to who produced different results.
-	status := NewChunkStatus(c, resultID, header.Height, agrees, disagrees)
+	// adds chunk status as a pending chunk to mempool.
+	status := &verification.ChunkStatus{
+		Chunk:             chunk,
+		ExecutionResultID: locator.ResultID,
+	}
 	added := e.pendingChunks.Add(status)
 	if !added {
-		return nil
-	}
-
-	allExecutors, err := e.state.Final().Identities(filter.HasRole(flow.RoleExecution))
-	if err != nil {
-		return fmt.Errorf("could not find executors: %w", err)
+		// chunk locators are deduplicated by consumer, reaching this point hints failing deduplication on consumer.
+		e.chunkConsumerNotifier.Notify(chunkID) // tells consumer that we are done with this chunk.
```

Comment:
Yes, it is fine and indeed is expected. `ChunkConsumer` pushes each chunk _locator_ on a separate Worker `goroutine`. If duplicate chunk _locators_ are pushed by `ChunkConsumer` to `fetcher` engine, it means that each chunk _locator_ arrives on a separate Worker. The `pendingChunks.Add` method is atomic and threadsafe, and hence it returning `false` means there is a duplicate chunk _locator_ already available. So we should spare the current Worker, and we do that by calling the notifier. 

Please note that we are talking about the chunk _locators_ and not the _chunk data packs_. A chunk _locator_ is the input ticket to the `fetcher` engine to retrieve the chunk, request its chunk data pack, and process it. We yet have not got any external and _unverified_ data object that needs verification. The verification happens when the chunk data pack arrives, which is the logic of another method of `fetcher` engine, i.e., [`HandleChunkDataPack`](https://github.com/onflow/flow-go/blob/348393948b601b285d631957bba625f13ca25c41/engine/verification/fetcher/engine.go#L180-L237).

Also, please note that we keep the Worker of a requested chunk data pack alive until either its chunk data arrives and [verification on it is conducted](https://github.com/onflow/flow-go/blob/348393948b601b285d631957bba625f13ca25c41/engine/verification/fetcher/engine.go#L229-L236), or the [requester notifies us that its corresponding block has been sealed](https://github.com/onflow/flow-go/blob/348393948b601b285d631957bba625f13ca25c41/engine/verification/fetcher/engine.go#L291-L296) and we no longer needed to verify it.

Reply:
Removing this line might cause the worker to hang forever.

But I wonder if this case could happen, I think we should just Fatal here, I can't think of a case where `consumer` would fail to dedup.

consumer deduplicates on chunk locator's ID, which is a hash of result ID and chunk index. If chunk locators are different, chunk ID must be different. Therefore, there won't be duplicates.

Reply:
Addressed in https://github.com/onflow/flow-go/pull/633/commits/713f586ba1fac58f1d4ef364baddb1bb712d0f0e